src/main/java/org/opensearch/knn/index/KNNIndexCache.java

```diff
@@ -312,38 +310,122 @@ public KNNIndexCacheEntry loadIndex(String indexPathUrl, String indexName) throw
         // the entry
         fileWatcher.init();
 
-        final KNNIndex knnIndex = KNNIndex.loadIndex(indexPathUrl, getQueryParams(indexName),
-                SpaceType.getSpace(KNNSettings.getSpaceType(indexName)).getValue());
+        // loadIndex from different library
+        String engineName;
+        Map<String, Object> parameters;
+        if (indexPathUrl.endsWith(KNNEngine.NMSLIB.getExtension())
+                || indexPathUrl.endsWith(KNNEngine.NMSLIB.getCompoundExtension())) {
+            engineName = KNNEngine.NMSLIB.getName();
+            parameters = ImmutableMap.of(
+                    KNNConstants.SPACE_TYPE, spaceType.getValue(),
+                    KNNConstants.HNSW_ALGO_EF_SEARCH, KNNSettings.getEfSearchParam(indexName)
+            );
+        } else {
+            throw new IllegalArgumentException("[KNN] Invalid engine type for path: " + indexPathUrl);
+        }
+
+        final long indexPointer = JNIService.loadIndex(indexPathUrl, parameters, engineName);
```

Comment:
may be indexReference? i feel pointer is more of cpp jargon.

Reply:
I use pointer here because it is generated in the C++ layer and used to access meemory exclusively in C++ layer. So, the value it contains is in fact a valid pointer in C++ and cannot be used to access memory in Java.